File: xy/internal/chgk/typo/typo.go
// Package typo is a Go port of chgksuite's typotools.py — the typography pass
// the parser applies to every field it extracts (quotes, dashes, stress accents,
// percent-decoding) plus the URL-aware underscore escaping the docx reader needs.
//
// Only the modes chgksuite's DefaultArgs use ("on") are implemented; the "smart"
// variants are reachable only from the command line and are treated as "off".
package typo

import (
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"
)

// ── whitespace ──────────────────────────────────────────────────────────────
//
// typotools.WHITESPACE is {space, NBSP, newline} — note it deliberately excludes
// tabs and carriage returns.

var (
	reBadWspStart = regexp.MustCompile(`^[ \x{00a0}\n]+`)
	reBadWspEnd   = regexp.MustCompile(`[ \x{00a0}\n]+$`)
	reWspNewline  = regexp.MustCompile(`[\s\x{00a0}]+\n[\s\x{00a0}]+`)
)

// REW ports remove_excessive_whitespace.
func REW(s string) string {
	s = reBadWspStart.ReplaceAllString(s, "")
	s = reBadWspEnd.ReplaceAllString(s, "")
	return reWspNewline.ReplaceAllString(s, "\n")
}

// ── URL spans ───────────────────────────────────────────────────────────────

// reURL ports typotools.re_url. The original is a mangled copy of a well-known
// "find URLs" pattern whose trailing alternations can only ever match empty, so
// it reduces to «scheme-ish prefix followed by a run of non-space, non-bracket
// characters». Python's \s is Unicode-aware, hence the explicit \p{Z}.
var reURL = regexp.MustCompile(`(?:[a-z][\w-]+:(?:/{1,3}|[a-z0-9%])|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)[^\s\p{Z}()<>]*`)

type span struct{ start, end int }

// httpURLSpans ports _iter_http_url_spans: a hand-rolled scanner that, unlike
// reURL, lets a URL contain balanced parentheses and drops one trailing ,.;
func httpURLSpans(s string) []span {
	var out []span
	for i := 0; i < len(s); {
		if !strings.HasPrefix(s[i:], "http://") && !strings.HasPrefix(s[i:], "https://") {
			_, sz := utf8.DecodeRuneInString(s[i:])
			i += sz
			continue
		}
		j := i + 1
		bracket := 0
		for j < len(s) {
			r, sz := utf8.DecodeRuneInString(s[j:])
			if unicode.IsSpace(r) || (r == ')' && bracket == 0) {
				break
			}
			if r == '(' {
				bracket++
			} else if r == ')' && bracket > 0 {
				bracket--
			}
			j += sz
		}
		end := j
		if j > i {
			if last := s[j-1]; last == ',' || last == '.' || last == ';' {
				end = j - 1
			}
		}
		out = append(out, span{i, end})
		i = j
	}
	return out
}

// urlSpans ports iter_url_spans: the http scan merged with the reURL matches.
func urlSpans(s string) []span {
	spans := httpURLSpans(s)
	for _, m := range reURL.FindAllStringIndex(s, -1) {
		spans = append(spans, span{m[0], m[1]})
	}
	kept := spans[:0]
	for _, sp := range spans {
		if sp.start < sp.end {
			kept = append(kept, sp)
		}
	}
	spans = kept
	// sort by (start, end) — Python sorts the tuples
	for i := 1; i < len(spans); i++ {
		for j := i; j > 0 && (spans[j].start < spans[j-1].start ||
			(spans[j].start == spans[j-1].start && spans[j].end < spans[j-1].end)); j-- {
			spans[j], spans[j-1] = spans[j-1], spans[j]
		}
	}
	var merged []span
	for _, sp := range spans {
		if n := len(merged); n > 0 && sp.start <= merged[n-1].end {
			if sp.end > merged[n-1].end {
				merged[n-1].end = sp.end
			}
			continue
		}
		merged = append(merged, sp)
	}
	return merged
}

var reUnescapedUnderscore = regexp.MustCompile(`\\?_`)

// EscapeUnderscoresExceptURLs ports escape_underscores_except_urls: underscores
// are the 4s italic marker, so they must be backslash-escaped everywhere except
// inside URLs (where they are common and meaningful).
func EscapeUnderscoresExceptURLs(s string, skipEscaped bool) string {
	if !strings.Contains(s, "_") {
		return s
	}
	escape := func(seg string) string {
		if !skipEscaped {
			return strings.ReplaceAll(seg, "_", `\_`)
		}
		// (?<!\\)_ — only escape underscores not already escaped.
		return reUnescapedUnderscore.ReplaceAllStringFunc(seg, func(m string) string {
			if m == `\_` {
				return m
			}
			return `\_`
		})
	}
	var b strings.Builder
	last := 0
	for _, sp := range urlSpans(s) {
		b.WriteString(escape(s[last:sp.start]))
		b.WriteString(s[sp.start:sp.end])
		last = sp.end
	}
	b.WriteString(escape(s[last:]))
	return b.String()
}

// ── quotes ──────────────────────────────────────────────────────────────────

// quoteFixer ports typotools.QuoteFixer: it walks the string tracking nesting
// depth and rewrites every quote character to the Russian convention («» at odd
// levels, „“ at even ones). If the quotes don't balance it gives up and returns
// the input untouched.
type quoteFixer struct {
	src   []rune
	out   []rune
	level int
	last  map[int]rune
	marks map[int]mark
}

type mark struct {
	opening bool
	level   int
}

func isSpaceRune(r rune) bool { return r == ' ' || r == ' ' }

func (q *quoteFixer) prev(i int) (rune, bool) {
	if i == 0 {
		return 0, false
	}
	return q.out[i-1], true
}

func (q *quoteFixer) next(i int) (rune, bool) {
	if i+1 == len(q.out) {
		return 0, false
	}
	return q.out[i+1], true
}

func (q *quoteFixer) fix() string {
	for i := range q.out {
		switch q.out[i] {
		case '«', '„':
			q.level++
			q.marks[i] = mark{true, q.level}
			q.last[q.level] = q.out[i]
		case '»', '”':
			q.marks[i] = mark{false, q.level}
			q.level--
		case '"':
			p, hasPrev := q.prev(i)
			n, hasNext := q.next(i)
			openish := q.level == 0 || !hasPrev || (isSpaceRune(p) && hasNext && !isSpaceRune(n))
			if openish {
				q.level++
				q.marks[i] = mark{true, q.level}
				q.last[q.level] = '"'
			} else if q.last[q.level] == '"' {
				q.marks[i] = mark{false, q.level}
				q.level--
			} else {
				q.level++
				q.marks[i] = mark{true, q.level}
				q.last[q.level] = '"'
			}
		case '“':
			if q.last[q.level] == '„' {
				q.marks[i] = mark{false, q.level}
				q.level--
			} else {
				q.level++
				q.marks[i] = mark{true, q.level}
			}
		}
	}
	if q.level != 0 {
		return string(q.src)
	}
	for i, m := range q.marks {
		if m.opening {
			if m.level%2 != 0 {
				q.out[i] = '«'
			} else {
				q.out[i] = '„'
			}
		} else {
			if m.level%2 != 0 {
				q.out[i] = '»'
			} else {
				q.out[i] = '“'
			}
		}
	}
	return string(q.out)
}

var (
	reApostropheAfter  = regexp.MustCompile(`(\w)'`)
	reApostropheBefore = regexp.MustCompile(`'(\w)`)
)

// GetQuotesRight ports get_quotes_right.
func GetQuotesRight(s string) string {
	if strings.Contains(s, `"`) || (strings.Contains(s, "“") && !strings.Contains(s, "„")) {
		r := []rune(s)
		q := &quoteFixer{
			src:   r,
			out:   append([]rune(nil), r...),
			last:  map[int]rune{},
			marks: map[int]mark{},
		}
		s = q.fix()
	}
	s = reApostropheAfter.ReplaceAllString(s, "$1’")
	s = reApostropheBefore.ReplaceAllString(s, "‘$1")
	return s
}

// ── dashes ──────────────────────────────────────────────────────────────────

// GetDashesRight ports get_dashes_right: a run of hyphens flanked by whitespace
// becomes an em dash. RE2 has no lookaround, so the flanks are checked by hand.
func GetDashesRight(s string) string {
	var b strings.Builder
	rs := []rune(s)
	for i := 0; i < len(rs); {
		if rs[i] == '-' && i > 0 && unicode.IsSpace(rs[i-1]) {
			j := i
			for j < len(rs) && rs[j] == '-' {
				j++
			}
			if j < len(rs) && unicode.IsSpace(rs[j]) {
				b.WriteRune('—')
				i = j
				continue
			}
			for ; i < j; i++ {
				b.WriteRune(rs[i])
			}
			continue
		}
		b.WriteRune(rs[i])
		i++
	}
	return strings.ReplaceAll(b.String(), " – ", " — ")
}

// ── accents ─────────────────────────────────────────────────────────────────

const (
	lowerRU     = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"
	upperRU     = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
	combAcute   = '́'
	potentialAc = "АОУЫЭЯЕЮИ"
)

var (
	badBeginnings = map[string]bool{"Мак": true, "мак": true, "О'": true, "о’": true, "О’": true, "о'": true}
	accentsToFix  = map[rune]bool{'̀': true, '́': true, '́': true}
	// latin→cyrillic homoglyphs, both cases (typotools.LETTERS_MAPPING)
	lettersMapping = map[rune]rune{
		'a': 'а', 'e': 'е', 'y': 'у', 'o': 'о', 'u': 'и',
		'A': 'А', 'E': 'Е', 'Y': 'У', 'O': 'О', 'U': 'И',
	}
	reNotRussian = regexp.MustCompile(`[^` + lowerRU + upperRU + `]+`)
)

func isCyrillic(r rune) bool {
	return strings.ContainsRune(lowerRU, unicode.ToLower(r))
}

// detectAccent ports typotools.detect_accent: in a mixed-case word an interior
// capital vowel is the chgk convention for marking stress ("мОсква"), so it is
// lowercased and given a combining acute.
func detectAccent(s string) string {
	for _, word := range reNotRussian.Split(s, -1) {
		if word == "" || strings.ToUpper(word) == word || utf8.RuneCountInString(word) <= 1 {
			continue
		}
		w := []rune(word)
		for i := 1; i < len(w); i++ {
			if !strings.ContainsRune(potentialAc, w[i]) {
				continue
			}
			if badBeginnings[string(w[:i])] {
				continue
			}
			if i != 1 && unicode.IsUpper(w[i-1]) {
				continue
			}
			if i+1 != len(w) && unicode.IsUpper(w[i+1]) {
				continue
			}
			nw := append([]rune{}, w[:i]...)
			nw = append(nw, unicode.ToLower(w[i]), combAcute)
			nw = append(nw, w[i+1:]...)
			w = nw
			i++ // skip past the combining mark we just inserted
		}
		if nw := string(w); nw != word {
			if idx := strings.Index(s, word); idx >= 0 {
				s = s[:idx] + nw + s[idx+len(word):]
			}
		}
	}
	return s
}

// cyrLatCheckChar ports cyr_lat_check_char: a Latin letter wedged between
// Cyrillic ones and carrying a combining accent is a homoglyph typo.
func cyrLatCheckChar(i int, word []rune) string {
	char := word[i]
	if isCyrillic(char) {
		return ""
	}
	leftOK := i == 0 || isCyrillic(word[i-1]) || !unicode.IsLetter(word[i-1])
	rightOK := i == len(word)-1 || isCyrillic(word[i+1]) || !unicode.IsLetter(word[i+1])
	if !leftOK || !rightOK {
		return ""
	}
	norm := []rune(nfd(char))
	if len(norm) > 1 && norm[0] != char {
		if mapped, ok := lettersMapping[norm[0]]; ok && accentsToFix[norm[1]] {
			return string(mapped) + string(combAcute) + string(norm[2:])
		}
	}
	return ""
}

// CyrLatCheckWord ports cyr_lat_check_word; it returns "" when nothing changes.
func CyrLatCheckWord(word string) string {
	w := []rune(word)
	if len(w) == 1 {
		return ""
	}
	type rep struct{ from, to string }
	var reps []rep
	for i := range w {
		if fixed := cyrLatCheckChar(i, w); fixed != "" {
			reps = append(reps, rep{string(w[i]), fixed})
		} else if isCyrillic(w[i]) && i < len(w)-1 && accentsToFix[w[i+1]] {
			reps = append(reps, rep{string(w[i]) + string(w[i+1]), string(w[i]) + string(combAcute)})
		}
	}
	if len(reps) == 0 {
		return ""
	}
	out := word
	seen := map[string]bool{}
	for _, r := range reps {
		if seen[r.from] {
			continue
		}
		seen[r.from] = true
		out = strings.ReplaceAll(out, r.from, r.to)
	}
	return out
}

func fixAccents(s string, on bool) string {
	if on {
		s = detectAccent(s)
	}
	type rep struct{ from, to string }
	var reps []rep
	seen := map[string]bool{}
	for _, word := range strings.Fields(s) {
		if seen[word] {
			continue
		}
		if fixed := CyrLatCheckWord(word); fixed != "" {
			seen[word] = true
			reps = append(reps, rep{word, fixed})
		}
	}
	for _, r := range reps {
		s = strings.ReplaceAll(s, r.from, r.to)
	}
	return s
}

// ── percent decoding ────────────────────────────────────────────────────────

var rePercent = regexp.MustCompile(`(%[0-9a-fA-F]{2})+`)

func unhex(b byte) byte {
	switch {
	case b >= '0' && b <= '9':
		return b - '0'
	case b >= 'a' && b <= 'f':
		return b - 'a' + 10
	default:
		return b - 'A' + 10
	}
}

// PercentDecode ports typotools.percent_decode: percent-escapes that decode to
// valid UTF-8 are turned back into text (chgk sources are full of pasted
// Wikipedia URLs). Longest runs first, so a shorter run can't clobber a longer.
func PercentDecode(s string) string {
	groups := rePercent.FindAllString(s, -1)
	// sort by length, descending (Python: sorted(..., key=len, reverse=True))
	for i := 1; i < len(groups); i++ {
		for j := i; j > 0 && len(groups[j]) > len(groups[j-1]); j-- {
			groups[j], groups[j-1] = groups[j-1], groups[j]
		}
	}
	for _, g := range groups {
		buf := make([]byte, 0, len(g)/3)
		for i := 0; i+2 < len(g); i += 3 {
			buf = append(buf, unhex(g[i+1])<<4|unhex(g[i+2]))
		}
		if !utf8.Valid(buf) {
			continue // Python raises and leaves the escape in place
		}
		s = strings.ReplaceAll(s, g, string(buf))
	}
	return s
}

// ── the pass itself ─────────────────────────────────────────────────────────

// Options mirrors the args.typography_* switches. The zero value is all-off;
// DefaultOptions is what chgksuite's DefaultArgs use.
type Options struct {
	Whitespace bool
	Quotes     bool
	Dashes     bool
	Accents    bool
	Percent    bool
}

// DefaultOptions is chgksuite's DefaultArgs: every knob "on".
func DefaultOptions() Options {
	return Options{Whitespace: true, Quotes: true, Dashes: true, Accents: true, Percent: true}
}

// Typography ports typotools.typography for the "on"/"off" modes.
func Typography(s string, o Options) string {
	if o.Whitespace {
		s = REW(s)
	}
	if o.Quotes {
		s = GetQuotesRight(s)
		s = strings.ReplaceAll(s, "'s", "’s")
	}
	if o.Dashes {
		s = GetDashesRight(s)
	}
	if o.Accents {
		s = fixAccents(s, true)
	}
	if o.Percent {
		s = PercentDecode(s)
	}
	return s
}

File: xy/web/ts/typo.ts
// typo.ts — the typography pass the card editor's "typograph" button and the
// board's "Typograph whole board" run: chgksuite's typotools (quotes -> guillemets,
// hyphen runs -> em dashes, percent-escapes -> the text they encode) plus the
// non-breaking-space/hyphen gluing, applied to 4s SOURCE.
//
// It runs in the browser because everything it touches is question text, and
// question text is exactly what the server must never see: a board-wide pass
// through an endpoint would have posted a whole package in the clear. The Go
// packages it was ported from (internal/chgk/typo + typoedit) stay where they
// are as the parity oracle — both suites read the same fixtures
// (internal/chgk/typoedit/testdata/pass_cases.json), so the two implementations
// cannot drift in silence.
//
// A port, quirks included: where chgksuite does something odd, so does this.
// Only the knobs the button promises are here — quotes, dashes and percent
// decoding. Whitespace normalisation would fight the editor, and accents have
// their own button (detect_accent rewrites any mixed-case word whether stress
// was meant or not).
//
// ONE KNOWN DIVERGENCE, older than this file and left alone deliberately: the
// non-breaking-hyphen rule. Go's inline.ReplaceNoBreak matches
// `[а-яё]{0,3}-[а-яё]{0,3}`, which — the bounds being zero — glues nearly every
// surviving hyphen; chgk.ts's replaceNoBreak requires `{1,3}` on both sides, so
// "raz- dva" keeps a plain hyphen here and gets a non-breaking one there. The Go
// side feeds the docx/PDF exporters, whose output is byte-parity-tested against
// chgksuite, so aligning them is a decision about which is right rather than a
// typo to fix. Until it is made, the fixtures deliberately avoid hyphens that
// fall in the gap.
//
// ES module.

import { xyChgk } from "./chgk.js";
import { xyVersions } from "./versions.js";

// ── quotes ──────────────────────────────────────────────────────────────────

// QuoteFixer walks the string tracking nesting depth and rewrites every quote to
// the Russian convention («» at odd levels, „“ at even). If the quotes do not
// balance it gives up and returns the input untouched — a half-fixed quotation
// is worse than none.
interface QuoteMark { opening: boolean; level: number }

function isSpaceChar(c: string): boolean { return c === " " || c === "\u00a0"; }

function fixQuotes(src: string): string {
  const out = [...src];
  const marks = new Map<number, QuoteMark>();
  const last = new Map<number, string>();
  let level = 0;
  for (let i = 0; i < out.length; i++) {
    const c = out[i];
    if (c === "«" || c === "„") {
      level++;
      marks.set(i, { opening: true, level });
      last.set(level, c);
    } else if (c === "»" || c === "”") {
      marks.set(i, { opening: false, level });
      level--;
    } else if (c === '"') {
      const prev = i === 0 ? null : out[i - 1];
      const next = i + 1 === out.length ? null : out[i + 1];
      const openish = level === 0 || prev === null ||
        (isSpaceChar(prev) && next !== null && !isSpaceChar(next));
      if (openish || last.get(level) !== '"') {
        level++;
        marks.set(i, { opening: true, level });
        last.set(level, '"');
      } else {
        marks.set(i, { opening: false, level });
        level--;
      }
    } else if (c === "“") {
      if (last.get(level) === "„") {
        marks.set(i, { opening: false, level });
        level--;
      } else {
        level++;
        marks.set(i, { opening: true, level });
      }
    }
  }
  if (level !== 0) return src;
  for (const [i, m] of marks) {
    out[i] = m.opening ? (m.level % 2 !== 0 ? "«" : "„") : (m.level % 2 !== 0 ? "»" : "“");
  }
  return out.join("");
}

function getQuotesRight(s: string): string {
  if (s.includes('"') || (s.includes("“") && !s.includes("„"))) s = fixQuotes(s);
  // JS \w is ASCII-only, like Go's RE2 and unlike Python's re.UNICODE: without
  // the class spelled out a Ukrainian apostrophe (Slov'yans'ka) keeps its quote.
  return s.replace(/([\p{L}\p{N}_])'/gu, "$1’").replace(/'([\p{L}\p{N}_])/gu, "‘$1");
}

// ── dashes ──────────────────────────────────────────────────────────────────

// Go's unicode.IsSpace, spelled out. JS's \s is NOT the same set: it omits
// U+0085 (NEL) and includes U+FEFF, so using it here would disagree with the
// oracle on both characters.
const RE_GO_SPACE = /[\t\n\v\f\r \u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]/;
const RE_GO_SPACE_RUN = /[\t\n\v\f\r \u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]+/;

// A run of hyphens flanked by whitespace becomes an em dash. Anything else — a
// hyphenated word, a list item's leading "-" (which never reaches here, see
// pass) — is left alone.
function getDashesRight(s: string): string {
  const rs = [...s];
  let out = "";
  for (let i = 0; i < rs.length;) {
    if (rs[i] === "-" && i > 0 && RE_GO_SPACE.test(rs[i - 1])) {
      let j = i;
      while (j < rs.length && rs[j] === "-") j++;
      if (j < rs.length && RE_GO_SPACE.test(rs[j])) {
        out += "—";
        i = j;
        continue;
      }
      for (; i < j; i++) out += rs[i];
      continue;
    }
    out += rs[i];
    i++;
  }
  return out.split(" – ").join(" — ");
}

// ── percent decoding ────────────────────────────────────────────────────────

const RE_PERCENT = /(?:%[0-9a-fA-F]{2})+/g;
// ignoreBOM keeps a decoded U+FEFF as a character instead of eating it, which is
// what Go's utf8.Valid + string() does: %EF%BB%BF decodes to the BOM, not to "".
const utf8 = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true });

// Percent-escapes that decode to valid UTF-8 become the text they stand for —
// chgk sources are made of pasted Wikipedia links. Longest runs first, so a
// shorter run cannot clobber a longer one it sits inside.
function percentDecode(s: string): string {
  const groups = (s.match(RE_PERCENT) || []).slice().sort((a, b) => b.length - a.length);
  for (const g of groups) {
    const bytes = new Uint8Array(g.length / 3);
    for (let i = 0; i * 3 < g.length; i++) bytes[i] = parseInt(g.slice(i * 3 + 1, i * 3 + 3), 16);
    let text: string;
    try {
      text = utf8.decode(bytes);
    } catch {
      continue; // not UTF-8: leave the escape exactly as it was
    }
    s = s.split(g).join(text);
  }
  return s;
}

// ── stress accents ──────────────────────────────────────────────────────────
// chgk marks stress by capitalising the vowel ("brAzer"), and detect_accent
// turns that into a real combining acute ("brazer" + acute). It is a HEURISTIC on
// capitalisation, so its guards are the whole substance: an all-caps word is a
// word in caps, not a stressed one; a capital next to another capital is part of
// a run; "Mac..." and "O'..." are name prefixes, not stress.
const COMB_ACUTE = "\u0301";
const LOWER_RU = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя";
const UPPER_RU = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ";
const POTENTIAL_ACCENT = "АОУЫЭЯЕЮИ";
const BAD_BEGINNINGS = new Set(["Мак", "мак", "О'", "о’", "О’", "о'"]);
const RE_NOT_RUSSIAN = new RegExp("[^" + LOWER_RU + UPPER_RU + "]+");
const ACCENTS_TO_FIX = new Set(["\u0300", "\u0341", "\u0301"]);
const LETTERS_MAPPING: Record<string, string> = {
  a: "а", e: "е", y: "у", o: "о", u: "и",
  A: "А", E: "Е", Y: "У", O: "О", U: "И",
};

const isUpper = (c: string): boolean => /\p{Lu}/u.test(c);
const isLetter = (c: string): boolean => /\p{L}/u.test(c);
const isCyrillic = (c: string): boolean => LOWER_RU.includes(c.toLowerCase());

// One word the heuristic would accent, as the board's review list shows it.
export interface AccentPick { from: string; to: string }

// `seen` collects what the pass WOULD do without doing it; `allow` restricts it
// to the pairs the editor ticked. Both are undefined for a plain pass, which is
// then chgksuite's own function exactly — the parity fixtures run that way.
// Reviewing is xy's, not chgksuite's, so the algorithm is untouched: this only
// watches it and, at the last moment, declines.
function detectAccent(s: string, seen?: Map<string, AccentPick>, allow?: Set<string>): string {
  for (const word of s.split(RE_NOT_RUSSIAN)) {
    if (word === "" || word.toUpperCase() === word || [...word].length <= 1) continue;
    let w = [...word];
    for (let i = 1; i < w.length; i++) {
      if (!POTENTIAL_ACCENT.includes(w[i])) continue;
      if (BAD_BEGINNINGS.has(w.slice(0, i).join(""))) continue;
      if (i !== 1 && isUpper(w[i - 1])) continue;
      if (i + 1 !== w.length && isUpper(w[i + 1])) continue;
      w = [...w.slice(0, i), w[i].toLowerCase(), COMB_ACUTE, ...w.slice(i + 1)];
      i++; // step over the combining mark just inserted
    }
    const nw = w.join("");
    if (nw === word) continue;
    if (seen) seen.set(word, { from: word, to: nw });
    if (allow && !allow.has(word)) continue;
    const idx = s.indexOf(word);
    if (idx >= 0) s = s.slice(0, idx) + nw + s.slice(idx + word.length);
  }
  return s;
}

// cyrLatCheckChar: a Latin letter wedged between Cyrillic ones and carrying a
// combining accent is a homoglyph typo — "moskva" typed with a Latin "o".
function cyrLatCheckChar(i: number, word: string[]): string {
  const char = word[i];
  if (isCyrillic(char)) return "";
  const leftOK = i === 0 || isCyrillic(word[i - 1]) || !isLetter(word[i - 1]);
  const rightOK = i === word.length - 1 || isCyrillic(word[i + 1]) || !isLetter(word[i + 1]);
  if (!leftOK || !rightOK) return "";
  const norm = [...char.normalize("NFD")];
  if (norm.length > 1 && norm[0] !== char) {
    const mapped = LETTERS_MAPPING[norm[0]];
    if (mapped && ACCENTS_TO_FIX.has(norm[1])) return mapped + COMB_ACUTE + norm.slice(2).join("");
  }
  return "";
}

function cyrLatCheckWord(word: string): string {
  const w = [...word];
  if (w.length === 1) return "";
  const reps: Array<[string, string]> = [];
  for (let i = 0; i < w.length; i++) {
    const fixed = cyrLatCheckChar(i, w);
    if (fixed !== "") reps.push([w[i], fixed]);
    else if (isCyrillic(w[i]) && i < w.length - 1 && ACCENTS_TO_FIX.has(w[i + 1])) {
      reps.push([w[i] + w[i + 1], w[i] + COMB_ACUTE]);
    }
  }
  if (!reps.length) return "";
  let out = word;
  const seen = new Set<string>();
  for (const [from, to] of reps) {
    if (seen.has(from)) continue;
    seen.add(from);
    out = out.split(from).join(to);
  }
  return out;
}

function fixAccents(s: string, seen?: Map<string, AccentPick>, allow?: Set<string>): string {
  s = detectAccent(s, seen, allow);
  const reps: Array<[string, string]> = [];
  const done = new Set<string>();
  for (const word of s.split(RE_GO_SPACE_RUN).filter((x) => x !== "")) {
    if (done.has(word)) continue;
    const fixed = cyrLatCheckWord(word);
    if (fixed !== "") {
      done.add(word);
      reps.push([word, fixed]);
    }
  }
  for (const [from, to] of reps) s = s.split(from).join(to);
  return s;
}

// ── the pass ────────────────────────────────────────────────────────────────

// How the caller wants stress marks handled: off, on, watched (`seen` fills with
// what it would do), or restricted to the words in `allow`.
export interface AccentOpts { seen?: Map<string, AccentPick>; allow?: Set<string> }

function typography(s: string, accents = false, o: AccentOpts = {}): string {
  s = getQuotesRight(s);
  s = s.split("'s").join("’s");
  s = getDashesRight(s);
  if (accents) s = fixAccents(s, o.seen, o.allow);
  return percentDecode(s);
}

// pass typographs 4s source, marker by marker. Splitting each line at its marker
// first is the whole point: a pass let loose on raw source reads a list item's
// leading "-" as a stray hyphen and turns it into an em dash, eating the list.
// Idempotent — the gluing rules match plain spaces, so text that already carries
// its NBSPs is left alone, and a button is something a user presses twice.
export function pass(source: string | null | undefined, accents = false, o: AccentOpts = {}): string {
  const lines = (source || "").split("\n");
  return lines.map((line) => {
    const { prefix, rest } = xyChgk.splitMarker(line);
    if (rest.trim() === "") return line;
    return prefix + xyChgk.replaceNoBreak(typography(rest, accents, o));
  }).join("\n");
}

// passAccents is pass plus stress-mark detection — what the buttons run.
export function passAccents(source: string | null | undefined, o: AccentOpts = {}): string {
  return pass(source, true, o);
}

// accentPicks reports every word the pass WOULD accent in these texts, without
// touching them: the board's review list, where an editor drops the compound
// nouns the heuristic cannot tell from a stress mark ("GazpromInvest"). Keyed by
// the original word, so the same word across thirty cards is one decision.
export function accentPicks(texts: ReadonlyArray<string>): AccentPick[] {
  const seen = new Map<string, AccentPick>();
  for (const t of texts) passAccents(t, { seen });
  return [...seen.values()];
}

// passVersions typographs a whole card — every version of it — and reassembles.
// The separators are xy's own markup, not prose, so they are never handed to the
// pass; and a wording the editor is not looking at needs straightening just as
// much as the one it is.
export function passVersions(desc: string | null | undefined, o: AccentOpts = {}): string {
  let out = desc || "";
  xyVersions.splitVersions(out).forEach((body, i) => {
    out = xyVersions.setVersionBody(out, i, passAccents(body, o));
  });
  return out;
}

export const xyTypo = {
  pass, passAccents, passVersions, accentPicks, typography,
  getQuotesRight, getDashesRight, percentDecode,
};
